### PR TITLE
De-personalize Diagnostic File URLs in Compiled Schemas

### PR DESCRIPTION
**Pull request by olabusayoT**
- update DFDLSchemaFile.uriString to remove any prefix from Classpath and replace $HOME with ~ using Misc helper function
- update systemId in SDE/SDW to replace $HOME with ~ using Misc helper function
- update/add tests

DAFFODIL-2195

### CODE REVIEW

File: daffodil-core/src/main/scala/org/apache/daffodil/core/dsom/DFDLSchemaFile.scala [hunk 1]
```diff
@@ -137,9 +145,29 @@ final class DFDLSchemaFile(
...
+    finalUriString
+  }
 
-  override protected lazy val diagnosticDebugNameImpl = schemaSource.uriForLoading.toString
+  override protected lazy val diagnosticDebugNameImpl = {
+    val uriString_ = schemaSource.uriForLoading.toString
+    val finalUriString = Misc.depersonalizePath(
+      uriString_
+        .replace(
+          Misc.classPath.find(cp => uriString_.startsWith(cp.toString)).getOrElse("").toString,
```
Comment (mbeckerle) [inline daffodil-core/src/main/scala/org/apache/daffodil/core/dsom/DFDLSchemaFile.scala:165]: This code is confusing. 

Can you spread this code out more? Add an intermediate for the uriString where the prefix that is found on the classpath has been removed, then depersonalize that in a separate line? The cascade of the replace with the find with the getOrElse is not easily understood.

File: daffodil-lib/src/main/scala/org/apache/daffodil/lib/api/Diagnostic.scala [hunk 2]
```diff
@@ -263,7 +263,12 @@ trait LocationInSchemaFile {
   ): String = {
     // works for both windows and unix
     val splitter = "/"
-    val tokens = uriString.split(splitter)
+    val classpathTrimmedUriString =
+      uriString.replace(
+        Misc.classPath.find(cp => uriString.startsWith(cp.toString)).getOrElse("").toString,
```
Comment (mbeckerle) [inline daffodil-lib/src/main/scala/org/apache/daffodil/lib/api/Diagnostic.scala:268]: Same idiom. I found confusing above.

Suggest create a private helper method and call it in both places. That way you can explain it once. 

Name it "stripClasspathPrefix" or something like that perhaps.

**Review by mbeckerle — CHANGES_REQUESTED** [review]
Small refactoring of some confusing code please.

File: daffodil-cli/src/test/scala/org/apache/daffodil/cli/cliTest/TestCLISaveParser.scala [hunk 3]
```diff
@@ -318,4 +318,18 @@ class TestCLISaveParser {
...
+  @Test def test_CLI_Saving_SaveParser_error(): Unit = {
+    val schema = path(
+      "daffodil-sapi/src/test/resources/test/sapi/mySchema6.dfdl.xsd",
+    )
+
+    withTempFile { parser =>
+      runCLI(args"save-parser -s $schema $parser") { cli =>
+        cli.expectErr("[error]")
+        cli.expectErr("systemId: file:~")
+        cli.expectErr("Schema context: file:~")
```
Comment (stevedlawrence) [inline daffodil-cli/src/test/scala/org/apache/daffodil/cli/cliTest/TestCLISaveParser.scala:330]: Instead of outputting `file:~/...`, it feels like this wants to be something like:
```
Schema context: daffodil-sapi/src/test/resources/test/sapi/mySchema6.dfdl.xsd",
```
since that is what the user provided. Similar to my other comment about how we want to use the impor/include schemaLocation as a hint of what to display, I think for the root schema we want to display the path that was actually provided.

As another example, if we had foo.jar on DAFFODIL_CLASSPATH and ran something like:
```
daffodil save-parser -s /some/file/in/foo/jar/dfdl.xsd
```
And that resolves to a path in foo.jar, then I think the schema context we output wants to be exactly `/some/file/in/foo/jar/dfdl.xsd`.

So I suggested that based on my other comment, when we resolve a schema, we either want to use the path directly (in the case of importing absolute schemaLocations or locations without an importing schema like this one), or resolve the schemaLocation relative to the importing display location in the case of relative imports with a context.

**Review by stevedlawrence — COMMENTED** [review]
Some comments on a potentially different approach. Let me know your thoughts, as I haven't put a ton of thought into it.

File: daffodil-lib/src/main/scala/org/apache/daffodil/lib/util/Misc.scala [hunk 4]
```diff
@@ -660,4 +660,15 @@ object Misc {
     }
     Some(res)
   }
+
+  /**
+   * Replace HOME env var in Path with ~
+   */
+  def depersonalizePath(path: String): String = {
+    path.replace(
+      System.getenv("HOME"),
```
[the remark this answers is not in the document]
Comment (tuxji) [reply daffodil-lib/src/main/scala/org/apache/daffodil/lib/util/Misc.scala:669]: Steve's suggestion makes sense to me (so does Mike's).  Also, Steve's PR #1067 has made the import/include logic more consistent, so you'll want to redo this PR against the latest HEAD anyway when you have time to work on it again.

File: daffodil-lib/src/main/scala/org/apache/daffodil/lib/exceptions/SchemaFileLocatable.scala [hunk 5]
```diff
@@ -35,24 +32,42 @@ trait HasSchemaFileLocation extends LookupLocation {
...
+  ) = {
+    new SchemaFileLocation(
+      lineNumber,
+      columnNumber,
+      uriString,
+      diagnosticFilepath,
+      contextToString,
+      diagnosticDebugName,
+    )
+  }
```
Comment (stevedlawrence) [inline daffodil-lib/src/main/scala/org/apache/daffodil/lib/exceptions/SchemaFileLocatable.scala:61]: This essentially just makes the `SchemaFileLocation` constructor public. If we wanted to keep it that way we could create a custom `XercesSchmeaFileLocatable`. Then the places where we use this new constructor, we instead do something like
```scala
val xsfl = new XerexSchemaFileLocatable(xercesError, ....)
val schemaLocation = SchemaLocation(xsfl)
new SchemaDefinitionError(schemaLocation, ...)
```
```

**Review by stevedlawrence — CHANGES_REQUESTED** [review]

File: daffodil-core/src/main/scala/org/apache/daffodil/core/dsom/SchemaSet.scala [hunk 6]
```diff
@@ -139,6 +139,8 @@ final class SchemaSet private (
    */
   override lazy val uriString: String = schemaSource.uriForLoading.toString
 
+  override lazy val diagnosticFilepath: String = schemaSource.diagnosticFilepath
```
[the remark this answers is not in the document]
Comment (olabusayoT) [reply daffodil-core/src/main/scala/org/apache/daffodil/core/dsom/SchemaSet.scala:142]: According to the comment above uriString, the schemaSource is the "first schema document". So my guess is we're using the root document in the schema set.

I decided to add the diagnosticFilepath everywhere where we set a URIString, specifically for the test "Testnamespaces.test_long_chain_05", where we have nested imports, I noticed diagnosticFilepath was returning as the base file, not the associated imported file. I'm not sure if this particular placement fixed the issue, but I wanted to play it safe!

File: daffodil-core/src/main/scala/org/apache/daffodil/core/compiler/Compiler.scala [hunk 7]
```diff
@@ -325,7 +325,7 @@ class Compiler private (
     optRootName: Option[String] = None,
     optRootNamespace: Option[String] = None,
   ): ProcessorFactory = {
-    val source = URISchemaSource(file.toURI)
+    val source = URISchemaSource((file.getPath, file.toURI))
```
[the remark this answers is not in the document]
Comment (olabusayoT) [reply daffodil-core/src/main/scala/org/apache/daffodil/core/compiler/Compiler.scala:328]: I think with the SchemaOption case class, we can pass that in as a single argument rather than using the tuple

File: daffodil-core/src/main/scala/org/apache/daffodil/core/dsom/DFDLSchemaFile.scala [hunk 8]
```diff
@@ -45,16 +45,42 @@ class DFDLSchemaFileLoadErrorHandler(schemaFileLocation: SchemaFileLocation)
   private def loaderErrors = loaderErrors_
   private def loaderWarnings = loaderWarnings_
 
-  private def loaderSDEs: Seq[Diagnostic] = loaderErrors.map {
-    new SchemaDefinitionError(schemaFileLocation, "Error loading schema due to %s", _)
+  private def loaderSDEs: Seq[Diagnostic] = loaderErrors.map { err =>
+    val errMessage = err.getMessage
```
[the remark this answers is not in the document]
Comment (olabusayoT) [reply daffodil-core/src/main/scala/org/apache/daffodil/core/dsom/DFDLSchemaFile.scala:49]: We need the schemaFileLocation for the diagnosticFilepath and uriString, so I don't think we can get rid of this. We don't use the file information from the Xerces error as it's a full file path

File: daffodil-core/src/main/scala/org/apache/daffodil/core/compiler/Compiler.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (stevedlawrence) [reply daffodil-core/src/main/scala/org/apache/daffodil/core/compiler/Compiler.scala:328]: The "SchemaOption" feels CLI specific to me and might not want to leak into the API.

[the remark this answers is not in the document]
Comment (olabusayoT) [reply daffodil-core/src/main/scala/org/apache/daffodil/core/compiler/Compiler.scala:328]: Hmm, any reason it can't be part of UriSchemaSource, that way it can be used widely?

[the remark this answers is not in the document]
Comment (stevedlawrence) [reply daffodil-core/src/main/scala/org/apache/daffodil/core/compiler/Compiler.scala:328]: I guess maybe it adds an extra level in indirection without much gain? For example, if `URISchemaSource` looked like this:
```scala
class URISchemaSource(val schemaOption)
```
Then to create it you need to create a separate object, e.g.
```scala
val schemaOption = new SchemaOption(..., ...)
val schemaSource = new URISchemaSource(schemaOption)
```

And then either everything needs to be changed to do `schemaSource.schemaOption.uri` for example, or `URISchemaSource` needs to update its members to reach into `schemaOption` for he uri and diagnostics.

Really, the only reason why we need `SchemaOption` is because scallop doesn't have a way to access to the original CLI argument after it's been converted. So you either need the a tuple (or a new object that's basically a tuple) to keep the original string around. Nothing else in the code really needs this pair of diagnostic  + uri separately/ And that's really all the URISchemaSource it's, it's a tuple of the URI and how to show that URI in diagnostics plus some helper stuff.

Which makes me wonder if instead of creating a `URI` in the `fileResourceURIConverter`, we should just create a `URISchemaSource` directly. I imagine everything that uses the `schema` option as a `URI` just creates a `URISchemaSource` anyways. And if it needs the `URI`, it can just access the `uri` member of the `URISchemaSource`?

File: daffodil-core/src/main/scala/org/apache/daffodil/core/dsom/DFDLSchemaFile.scala [hunk 9]
```diff
@@ -137,9 +163,16 @@ final class DFDLSchemaFile(
 
   override lazy val optXMLSchemaDocument = iiParent.optXMLSchemaDocument
 
-  override lazy val uriString = schemaSource.uriForLoading.toString
+  override lazy val uriString = {
+    val uriString_ = schemaSource.uriForLoading.toString
+    uriString_
+  }
```
[the remark this answers is not in the document]
Comment (olabusayoT) [reply daffodil-core/src/main/scala/org/apache/daffodil/core/dsom/DFDLSchemaFile.scala:169]: Reverted this change

**Review by mbeckerle — APPROVED** [review]
+1

I have no further objections other than the things Steve has raised that are unresolved as yet.

File: daffodil-core/src/main/scala/org/apache/daffodil/core/dsom/IIBase.scala [hunk 10]
```diff
@@ -195,15 +196,32 @@ abstract class IIBase(
...
             WarnID.DeprecatedRelativeSchemaLocation,
             relToAbs,
             s"Resolving relative schemaLocations absolutely is deprecated. Did you mean /$slText",
           )
-          URISchemaSource(uri)
+          // if isBootStrapSD is true, we assume we are using the fakeXMLSchemaDocument, which means
+          // we will be passing in and receiving back an absolute diagnosticFilepath from resolveSchemaLocation.
+          // In just this case, we want to ignore that absolute filepath and use the diagnosticFilepath
+          // from main, which is the XMLSchemaDocument diagnosticFilepath
+          val finalUriSchemaSource = if (xsdArg.isBootStrapSD) {
```
Comment (tuxji) [inline daffodil-core/src/main/scala/org/apache/daffodil/core/dsom/IIBase.scala:218]: What do the letters "SD" stand for in "isBootStrapSD"?

File: daffodil-japi/src/test/java/org/apache/daffodil/example/TestJavaAPI.java [hunk 11]
```diff
@@ -122,7 +122,7 @@ public void testJavaAPI1() throws IOException, ClassNotFoundException {
         DebuggerRunnerForJAPITest debugger = new DebuggerRunnerForJAPITest();
 
         org.apache.daffodil.japi.Compiler c = Daffodil.compiler();
-        java.io.File schemaFile = getResource("/test/japi/mySchema1.dfdl.xsd");
+        java.io.File schemaFile = new File("daffodil-japi/src/test/resources/test/japi/mySchema1.dfdl.xsd");
```
Comment (tuxji) [inline daffodil-japi/src/test/java/org/apache/daffodil/example/TestJavaAPI.java:125]: What is the reason for replacing getResource (was it a function we defined ourselves?  I thought Java's getResource returned an URI or null, not a File) with new File?

File: daffodil-lib/src/main/scala/org/apache/daffodil/lib/exceptions/SchemaFileLocatable.scala (hunk 5 above)
Comment (tuxji) [reply daffodil-lib/src/main/scala/org/apache/daffodil/lib/exceptions/SchemaFileLocatable.scala:56]: I think your comment is already resolved, Steve.  This PR defines a new custom XercesSchemaFileLocatable and passes it to this new apply() constructor.  The places where the apply() constructor is called already looks very similar to your code except for the addition of a second parameter.  The second parameter looks important and shouldn't be omitted.

```scala
    val xsfl = new XercesSchemaFileLocatable(err, "")
    val newSchemaLocation = SchemaFileLocation(
      xsfl,
      schemaFileLocation,
    )
    new SchemaDefinitionError(
      newSchemaLocation,
      "Error loading schema due to %s",
      errMessage,
    )
```

**Review by tuxji — APPROVED** [review]
+1

Looks good to me.

File: daffodil-lib/src/main/scala/org/apache/daffodil/lib/xml/XMLUtils.scala [hunk 12]
```diff
@@ -1488,8 +1499,24 @@ Differences were (path, expected, actual):
...
+              new File(contextURIDFP).toPath
+            } else {
+              Paths.get(Misc.stripJarPrefix(contextURIDFP))
+            }
+            val relativeDiagnosticFilepath =
+              if (new File(uri.getPath).compareTo(diagnosticFilepathPaths.toFile) == 0) { // paths are the same
+                diagnosticFilepathPaths.toString
+              } else {
+                diagnosticFilepathPaths.resolveSibling(uri.getPath).toString
+              }
```
Comment (stevedlawrence) [inline daffodil-lib/src/main/scala/org/apache/daffodil/lib/xml/XMLUtils.scala:1513]: We need some comments about what is going on here. I'm not sure I understand why we need this.

Seem this logic is because sometimes the diagnosticFilePath isn't a Path, like it's a URI or something? Maybe if we require that it must be a Path, mentioned in another comment, then this special case  goes away? If diagnosticFilePath is just a Path, we can always just call resolveSibling here and it should just work?

File: daffodil-tdml-lib/src/main/scala/org/apache/daffodil/tdml/TDMLRunner.scala [hunk 13]
```diff
@@ -758,7 +760,12 @@ abstract class TestCase(testCaseXML: NodeSeq, val parent: DFDLTestSuite) {
       case (None, Some(uri)) => {
         //
         // In this case, we have a real TDML file (or resource) to open
-        URISchemaSource(uri)
+        val path = if (Misc.isNullOrBlank(uri.getPath)) {
```
Comment (stevedlawrence) [inline daffodil-tdml-lib/src/main/scala/org/apache/daffodil/tdml/TDMLRunner.scala:763]: Is this in case uri is a "jar" URI? Maybe this goes away with the `uriToDiagnosticPath` idea mentioend above?

This probably raises an issue that we want to be careful about calling uri.getPath, since that doesn't work for "jar" uris. I imagine we don't do that except for converting URI's to diagnostics.

File: daffodil-test/src/test/resources/org/apache/daffodil/section00/general/disallowDocTypes.tdml [hunk 14]
```diff
@@ -59,8 +59,6 @@
     <tdml:document/>
     <tdml:errors>
       <tdml:error>Schema Definition Error</tdml:error>
-      <tdml:error>org.xml.sax.SAXParseException</tdml:error>
-      <tdml:error>hasDocType-include.dfdl.xsd</tdml:error>
       <tdml:error>DOCTYPE is disallowed</tdml:error>
```
Comment (stevedlawrence) [inline daffodil-test/src/test/resources/org/apache/daffodil/section00/general/disallowDocTypes.tdml:62]: hasDocType-include.dfdl.xsd looks like a diagnosticPath. Did we lose that?

File: daffodil-cli/src/test/scala/org/apache/daffodil/cli/cliTest/TestCLISaveParser.scala [hunk 15]
```diff
@@ -318,4 +318,20 @@ class TestCLISaveParser {
...
+      "daffodil-sapi/src/test/resources/test/sapi/mySchema6.dfdl.xsd",
+    )
+
+    withTempFile { parser =>
+      runCLI(args"save-parser -s $schema $parser") { cli =>
+        cli.expectErr("[error]")
+        cli.expectErr(s"Schema context:  Location line 32 column 74 in $schema")
+      }(ExitCode.UnableToCreateProcessor)
+    }
+  }
```
Comment (stevedlawrence) [inline daffodil-cli/src/test/scala/org/apache/daffodil/cli/cliTest/TestCLISaveParser.scala:335]: I manually ran this in a CLI and the diagostic location looks good, but I got the same error twice:

```
[error] Schema Definition Error: Error loading schema due to src-resolve: Cannot resolve the name 'tns:nonExistent' to a(n) 'type definition' component.
Schema context:  Location line 32 column 74 in daffodil-sapi/src/test/resources/test/sapi/mySchema6.dfdl.xsd
[error] Schema Definition Error: Error loading schema due to src-resolve: Cannot resolve the name 'tns:nonExistent' to a(n) 'type definition' component.
Schema context:  Location line 32 column 74 in daffodil-sapi/src/test/resources/test/sapi/mySchema6.dfdl.xsd
```

I ran the same command on the main branch and only get it once (but with a much less human readable error emssage). I'm not sure what would have changed to break that. But this error does come from xerces and this PR does slightly change how xerces errors are handled, maybe something is wrong there?

**Review by stevedlawrence — COMMENTED** [review]
Looks good! Feels like the right approach and I've doe some basic testing and it looks to be doing the right thing. Just some minor comments that might help simplify things a bit, and maybe remove some complex edge cases.

File: daffodil-core/src/main/scala/org/apache/daffodil/core/dsom/DFDLSchemaFile.scala [hunk 16]
```diff
@@ -45,16 +46,38 @@ class DFDLSchemaFileLoadErrorHandler(schemaFileLocation: SchemaFileLocation)
...
   private def loaderWarnings = loaderWarnings_
 
-  private def loaderSDEs: Seq[Diagnostic] = loaderErrors.map {
-    new SchemaDefinitionError(schemaFileLocation, "Error loading schema due to %s", _)
+  private def loaderSDEs: Seq[Diagnostic] = loaderErrors.map { err =>
+    val errMessage = err.getMessage
+    // we create a new SchemaFileLocation because the Xerces error has line and column info that
+    // the original schemaFileLocation that's passed in doesn't contain, so we create this more
+    // complete SchemaFileLocation
+    val xsfl = new XercesSchemaFileLocatable(err, "")
```
[the remark this answers is not in the document]
Comment (olabusayoT) [reply daffodil-core/src/main/scala/org/apache/daffodil/core/dsom/DFDLSchemaFile.scala:54]: That was intentional. It represents the diagnostic debug name, which we changed from "Schema File" to an empty string

[the remark this answers is not in the document]
Comment (stevedlawrence) [reply daffodil-core/src/main/scala/org/apache/daffodil/core/dsom/DFDLSchemaFile.scala:54]: I see, makes sense. So does it always want to be empty string then? Maybe instead of it being a parameter XercesSchemaFileLocatable can do something like  `override diagnosticDebugName =  ""` with a comment explaining why that's preferred over "Schema File"?

File: daffodil-test/src/test/resources/org/apache/daffodil/section00/general/disallowDocTypes.tdml (hunk 14 above)
Comment (olabusayoT) [reply daffodil-test/src/test/resources/org/apache/daffodil/section00/general/disallowDocTypes.tdml:62]: It's actually the systemId from the XercesError which we lost when we converted it to a schemaFileLocation since we don't grab the absolute system id from the Xerces error and use the diagnosticFilepath from the schemaFileLocation passed into the errorhandler instead

File: daffodil-lib/src/main/scala/org/apache/daffodil/lib/api/DaffodilSchemaSource.scala [hunk 17]
```diff
@@ -58,24 +58,29 @@ sealed trait DaffodilSchemaSource {
...
 object URISchemaSource {
-  def apply(fileOrResource: URI) = {
-    new URISchemaSource(fileOrResource)
+  def apply(fileOrResourceRaw: String, fileOrResourceUri: URI): URISchemaSource = {
+    new URISchemaSource(fileOrResourceRaw, fileOrResourceUri)
   }
 }
 
-class URISchemaSource protected (val fileOrResource: URI) extends DaffodilSchemaSource {
+class URISchemaSource protected (val raw: String, val uri: URI) extends DaffodilSchemaSource {
```
[the remark this answers is not in the document]
Comment (olabusayoT) [reply daffodil-lib/src/main/scala/org/apache/daffodil/lib/api/DaffodilSchemaSource.scala:74]: Uri.getPath returns a string, not a nio Path, but we can use the uriToDiagnosticPath function to ensure we return nio Paths instead of string

File: daffodil-lib/src/main/scala/org/apache/daffodil/lib/util/Misc.scala [hunk 18]
```diff
@@ -708,4 +708,19 @@ object Misc {
...
+   * @return
+   */
+  def stripJarPrefix(uriString: String): String = {
+    val withoutJar = if (!isNullOrBlank(uriString)) {
+      uriString.split("\\.jar!").last
+    } else {
+      uriString
+    }
+    withoutJar
+  }
```
[the remark this answers is not in the document]
Comment (olabusayoT) [reply daffodil-lib/src/main/scala/org/apache/daffodil/lib/util/Misc.scala:724]: In windows, we ran into some places where a jar uri had a "file" scheme, so I'm going to treat both schemes the same

[the remark this answers is not in the document]
Comment (stevedlawrence) [reply daffodil-lib/src/main/scala/org/apache/daffodil/lib/util/Misc.scala:724]: So the URI was something like `file://path/to/foo.jar!/foo.dfdl.xsd` but with windows paths? I don't think that is a valid file path, since anything opening that is going to look for a file called `foo.dfdl.xsd" in a directory called `foo.jar!`, I think. I wonder if we had a bug somewhere else that lead to that invalid file scheme URI?

File: daffodil-core/src/main/scala/org/apache/daffodil/core/dsom/SchemaComponentIncludesAndImportsMixin.scala [hunk 19]
```diff
@@ -35,4 +35,6 @@ trait SchemaComponentIncludesAndImportsMixin extends CommonContextMixin {
     xmlSchemaDocument.uriString
   }.toOption.getOrElse(orElseURL)
```
[the remark this answers is not in the document]
Comment (olabusayoT) [reply daffodil-core/src/main/scala/org/apache/daffodil/core/dsom/SchemaComponentIncludesAndImportsMixin.scala:36]: I don't think this trait goes away since a lot of SchemaComponents extend it, and it seems like a straightforward way to set diagnosticFilepath. I do think uriString can go away as we don't seem to be using it anymore

File: daffodil-lib/src/main/scala/org/apache/daffodil/lib/api/DaffodilSchemaSource.scala (hunk 17 above)
[the remark this answers is not in the document]
Comment (olabusayoT) [reply daffodil-lib/src/main/scala/org/apache/daffodil/lib/api/DaffodilSchemaSource.scala:74]: So it looks like java.nio.file.Path is not serializable and it causes a bunch of serialization errors in the tdmlprocessor. So I'm not sure we can make diagnosticPath a java.nio.file.Path

File: daffodil-lib/src/main/scala/org/apache/daffodil/lib/util/Misc.scala (hunk 18 above)
[the remark this answers is not in the document]
Comment (olabusayoT) [reply daffodil-lib/src/main/scala/org/apache/daffodil/lib/util/Misc.scala:724]: Ah I see. I'll treat them differently and reinvestigate why it does that in Windows

File: daffodil-core/src/main/scala/org/apache/daffodil/core/dsom/SchemaComponentIncludesAndImportsMixin.scala (hunk 19 above)
[the remark this answers is not in the document]
Comment (olabusayoT) [reply daffodil-core/src/main/scala/org/apache/daffodil/core/dsom/SchemaComponentIncludesAndImportsMixin.scala:36]: Just a note, schemaLocation.uriString is needed still. But the schemaComponents.uriString is not

File: daffodil-sapi/src/test/scala/org/apache/daffodil/example/TestScalaAPI.scala [hunk 20]
```diff
@@ -345,7 +345,9 @@ class TestScalaAPI {
   def testScalaAPI4b(): Unit = {
     val c = Daffodil.compiler()
 
-    val schemaFileName = getResource("/test/sapi/mySchema3.dfdl.xsd")
```
Comment (stevedlawrence) [inline daffodil-sapi/src/test/scala/org/apache/daffodil/example/TestScalaAPI.scala:348]: Should we revert these changes so they match the Java API tests?

File: daffodil-tdml-lib/src/main/scala/org/apache/daffodil/tdml/TDMLRunner.scala [hunk 21]
```diff
@@ -320,14 +321,22 @@ class DFDLTestSuite private[tdml] (
       Logger.log.info(s"loading TDML file: ${tdmlFile}")
       val uri = tdmlFile.toURI()
       val newNode =
-        loader.load(URISchemaSource(uri), optTDMLSchema, addPositionAttributes = true)
+        loader.load(
+          URISchemaSource(Misc.uriToDiagnosticFile(uri), uri),
```
Comment (stevedlawrence) [inline daffodil-tdml-lib/src/main/scala/org/apache/daffodil/tdml/TDMLRunner.scala:325]: Use the `file` for the diag.

**Review by stevedlawrence — APPROVED** [review]
+1, only suggestion is to avoid using `uriToDiagnosticFile`, that should really only be used when all we have is a URI and know nothing about where it came from. If we have more information, we can probably use that to make a better diagnosticFile.
